### DISCUSSION TITLE
AuthenticatedHttpRequest use for LoginRequiredMixin

### DISCUSSION THREAD

**Discussion by H4rryK4ne** [typeddjango/django-stubs#2046] [Q&A] [answered]
As described in https://github.com/typeddjango/django-stubs#how-can-i-create-a-httprequest-thats-guaranteed-to-have-an-authenticated-user one could use `AuthenticatedHttpRequest` annotate `request` in function views with `@login_required` decorators. 

But how can this be adapted to class views with `LoginRequiredMixin`?

Any idea?

**Comment by mschoettle** [comment]
You can change the type hint for the `request` argument to `AuthenticatedHttpRequest`.

```py
class SomeAPIView(...):
    def get(self, request: AuthenticatedHttpRequest, ...)
```

**Comment by H4rryK4ne** [comment]
Sorry for the late reply ... 

Finally I got to the point to refactor my app and to try your solution. But I get the following error message from mypy v1.8.0
```
my_app/views/__init__.py:70: error: Argument 1 of "get" is incompatible with supertype "ProcessFormView"; supertype defines the argument type as "HttpRequest"  [override]
        def get(self, request: AuthenticatedHttpRequest, *args: Any, **kwargs: Any) -> HttpResponse:
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

```python
class User(AbstractUser):
    ...

class AuthenticatedHttpRequest(HttpRequest):
    user: User
```

**Comment by mschoettle** [comment]
Indeed. My bad.

I unfortunately don't know any solution except explicitly `cast()`ing `request` to an `authenticated_request` in the method body.

**Reply by intgr** [reply]
Sorry, this thread became disjointed because I converted [the issue](https://github.com/typeddjango/django-stubs/issues/1886) to a discussion.

**Comment by intgr** [comment, marked as answer]
Personally I never liked the `AuthenticatedHttpRequest` hack.

I always use explicit `isinstance()` guards when I need to access `request.user`. This adds a little runtime overhead, but is safer because it always validates the assumptions you're making. For example:

```python
from django.contrib.auth.models import User

class SomeAPIView(...):
    def get(self, request: HttpRequest, ...):
        assert isinstance(request.user, User)  # Type guard
        print(request.user.email)              # No error because mypy now knows the type of `request.user`
```

**Reply by mschoettle** [reply]
This works as long as optimization is not requested: https://docs.astral.sh/ruff/rules/assert/

**Reply by intgr** [reply]
True. But it's no worse than the `AuthenticatedHttpRequest` work-around. When using `AuthenticatedHttpRequest`, there is no runtime checking at all.
